### PR TITLE
Add script to identify broken links

### PR DESCRIPTION
**Pull request by ckadner**
**What this PR does / why we need it**:

_**Motivation**_

- Broken links in documentation are making it harder for new-comers to learn about KServe
- It often is not easy to find what the correct link should be
- For project maintainers it is time consuming to revisit documentation to identify broken links manually

_**Proposed Changes**_

- Add new script `hack/verify-doc-links.py`
- Add new `make` target `verify-doc-links` 
- The script found 1245 links (533 unique URLs) in 187 Markdown files with 51 links broken
- I made a "best-effort" attempt to fix 37 of the 51 broken links identified by the new script
- There are 14 broken links remaining. I might need some help from the KServe community to address those

_**Example usage**_

```Bash
$ make check-doc-links
```
```YAML
Checked 1246 links (534 unique URLs) in 187 Markdown files.

README.md:3: https://coveralls.io/github/kserve/kserve?branch=master -> 404
docs/samples/README.md:36: https://github.com/kserve/kserve/tree/master/python/lightgbm -> 404
docs/samples/explanation/alibi/moviesentiment/README.md:371: /docs/samples/explanation/alibi/moviesentiment/./kfserving_text_explainer.ipynb -> 404
python/kserve/docs/V1alpha1ClusterServingRuntime.md:11: /python/kserve/docs/.md -> 404
python/kserve/docs/V1alpha1ModelSpec.md:8: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1alpha1ServingRuntime.md:11: /python/kserve/docs/.md -> 404
python/kserve/docs/V1alpha1TrainedModel.md:11: /python/kserve/docs/V1alpha1TrainedModelStatus.md -> 404
python/kserve/docs/V1beta1CustomExplainer.md:24: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1CustomPredictor.md:24: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1CustomTransformer.md:24: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1ExplainerSpec.md:33: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1PodSpec.md:24: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1PredictorSpec.md:33: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1TransformerSpec.md:30: /python/kserve/docs/ResourceQuantity.md -> 404

ERROR: Found 14 invalid Markdown links
```

_**How does the script work?**_

* Find all Markdown files in the project
* Extract all Markdown-style links and plain URLs
* Create unique list of simplified URLs
* Use local files to check relative links, use cached HTTP requests for links outside project
* Run HTTP requests in parallel
* Retry GitHub rate-limited URLs after (at least) one minute
* Verify YouTube videos are actually still available (Youtube will return `200` even for unavailable videos)
* Report broken links with source file and line number

**_Script pedigree_**

- I contributed several iteration of this script to other projects I worked on before:
  - [MLX](https://github.com/machine-learning-exchange/mlx/blob/main/tools/python/verify_doc_links.py)
  - [KFP-Tekton](https://github.com/kubeflow/kfp-tekton/blob/master/tools/python/verify_doc_links.py)
  - [ModelMesh](https://github.com/kserve/modelmesh-serving/pull/237)
- This latest one differs from earlier version in that it does not require any non-standard libraries like for Markdown or `requests` to be installed into a virtual environment prior to running the script and it verifies YouTube videos are still available.

_**Possible future enhancements**_

- Add a GitHub action to run the `make check-doc-links` with any new PR
- This could help to identify broken links as soon as possible
- Hopefully the PR contributor (and reviewers) could then fix the broken link(s) -- even if outside the scope of the PR.

**Type of changes**

- [x] Developer tools (tools and scripts that help KServe developers)
- [x] This change requires a documentation update

**Feature/Issue validation/testing**:

- There are no functional code changes in any source files modified.

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
NONE
```

---

/cc @njhill

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # KServe
-[![go.dev reference](https://img.shields.io/badge/go.dev-reference-007d9c?logo=go&logoColor=white)](https://pkg.go.dev/github.com/kserve/kserve)
+[![go.dev reference](https://img.shields.io/badge/go.devil-reference-007d9c?logo=go&logoColor=white)](https://pkg.go.dev/github.com/kserve/kserve)
```
Comment (yuzisun) [inline README.md]: what's the difference between `go.dev` and `go.devil` ?

File: python/kserve/docs/V1alpha1ClusterServingRuntime.md [hunk 2]
```diff
@@ -2,13 +2,13 @@
...
+**kind** | **str**                                                                                                     | Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds | [optional] 
 **metadata** | [**V1ObjectMeta**](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1ObjectMeta.md) |  | [optional] 
-**spec** | [**V1alpha1ServingRuntimeSpec**](V1alpha1ServingRuntimeSpec.md) |  | [optional] 
-**status** | [**object**](.md) | ServingRuntimeStatus defines the observed state of ServingRuntime | [optional] 
+**spec** | [**V1alpha1ServingRuntimeSpec**](V1alpha1ServingRuntimeSpec.md)                                             |  | [optional] 
+**status** | [**V1alpha1ServingRuntimeStatus**](V1alpha1ServingRuntimeStatus.md)                                                                     | ServingRuntimeStatus defines the observed state of ServingRuntime | [optional]
```
Comment (yuzisun) [inline python/kserve/docs/V1alpha1ClusterServingRuntime.md]: Do we actually have status for `ServingRuntime` ?

File: python/kserve/docs/V1beta1ExplainerSpec.md [hunk 3]
```diff
@@ -30,13 +30,13 @@ Name | Type | Description | Notes
...
+**overhead** | [**dict(str, ResourceQuantity)**](ResourceQuantity.md) | Overhead represents the resource overhead associated with running a pod for a given RuntimeClass. This field will be autopopulated at admission time by the RuntimeClass admission controller. If the RuntimeClass admission controller is enabled, overhead must not be set in Pod create requests. The RuntimeClass admission controller will reject Pod create requests which have the overhead already set. If RuntimeClass is configured and selected in the PodSpec, Overhead will be set to the value defined in the corresponding RuntimeClass, otherwise it will remain unset and treated as zero. More info: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/688-pod-overhead/README.md This field is alpha-level as of Kubernetes v1.16, and is only honored by servers that enable the PodOverhead feature. | [optional]
```
Comment (yuzisun) [inline python/kserve/docs/V1beta1ExplainerSpec.md]: These are generated from the [PodSpec](https://github.com/kserve/kserve/blob/master/pkg/apis/serving/v1beta1/podspec.go#L232), you did regenerate the doc from swagger.json right?

File: README.md (hunk 1 above)
Comment (ckadner) [reply README.md]: Oh, good catch. That was me testing if I could produce an error on changing the image shield URL to see if the script could catch it. Turns out it returns `200` anyway. I will revert that right away.

File: python/kserve/docs/V1alpha1ClusterServingRuntime.md (hunk 2 above)
Comment (ckadner) [reply python/kserve/docs/V1alpha1ClusterServingRuntime.md]: No, that was one of the links I could not resolve, like the other 13 listed in the PR description that I need help with :-)

File: python/kserve/docs/V1beta1ExplainerSpec.md (hunk 3 above)
Comment (ckadner) [reply python/kserve/docs/V1beta1ExplainerSpec.md]: I was only trying to fix as many of the links as I could find. I am new to KServe so I am not yet familiar with the codegen process.

The intention of this PR is to provide a tool to identify broken links. I don't have the knowledge to fix them all :-)

Comment (yuzisun) [reply python/kserve/docs/V1beta1ExplainerSpec.md]: Thanks for the great tool!! Actually I just bumped up the kube version a few days ago and that seems like fixed links here so we could just run the [codegen](https://github.com/kserve/kserve/tree/master/hack/python-sdk#generate-python-sdk) to update the docs.

**Comment by ckadner** [comment]
Hi Dan, thank you for starting to review my PR!

The intention of this PR is to provide a tool to identify broken links. I did not have the expectation that I could also fix all of the broken links as I am new to KServe.

This was the tally of broken links before I attempted to fix any of them:

```YAML
Checked 1245 links (533 unique URLs) in 187 Markdown files.

README.md:3: https://coveralls.io/github/kserve/kserve?branch=master -> 404
docs/guidelines/kubeflow_kfserving_requirements.md:38: https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/ -> 404
docs/guidelines/kubeflow_kfserving_requirements.md:38: https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#specify-multiple-versions -> 404
docs/guidelines/kubeflow_kfserving_requirements.md:41: https://github.com/kubeflow/common/blob/master/job_controller/api/v1/types.go -> 404
docs/guidelines/kubeflow_kfserving_requirements.md:91: https://github.com/cncf/toc/blob/master/process/graduation_criteria.adoc -> 404
docs/samples/README.md:36: https://github.com/kserve/kserve/tree/master/python/lightgbm -> 404
docs/samples/autoscaling/README.md:27: https://knative.dev/docs/serving/installing-logging-metrics-traces -> 404
docs/samples/autoscaling/README.md:220: https://github.com/knative/serving/blob/master/docs/scaling/DEVELOPMENT.md -> 404
docs/samples/autoscaling/README.md:220: https://github.com/knative/serving/blob/master/docs/scaling/DEVELOPMENT.md#panic-mode -> 404
docs/samples/bentoml/README.md:6: https://docs.bentoml.org/en/latest/examples.html -> 404
docs/samples/bentoml/README.md:184: https://docs.bentoml.org/en/latest/quickstart.html -> 404
docs/samples/custom-domain/README.md:76: https://github.com/kubeflow/kfserving/tree/master/docs/samples/sklearn -> 404
docs/samples/explanation/alibi/moviesentiment/README.md:371: /docs/samples/explanation/alibi/moviesentiment/./kfserving_text_explainer.ipynb -> 404
docs/samples/gcp-iap/README.md:151: https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/_auth.py -> 404
docs/samples/istio-dex/README.md:5: /docs/samples/istio-dex/../sklearn -> 404
docs/samples/pipelines/README.md:16: /docs/samples/pipelines/./kfs-pipeline-v1alpha2.ipynb -> 404
docs/samples/storage/s3/README.md:5: https://knative.dev/docs/serving/outbound-network-access/ -> 404
docs/samples/storage/storageSpec/README.md:5: https://knative.dev/docs/serving/outbound-network-access/ -> 404
docs/samples/storage/uri/README.md:9: https://knative.dev/docs/serving/outbound-network-access/ -> 404
docs/samples/v1beta1/custom/torchserve/docs/canary.md:104: https://knative.dev/docs/install/install-kn/ -> 404
docs/samples/v1beta1/lightgbm/README.md:103: /docs/samples/v1beta1/lightgbm/../../../python/lgbserver/README.md -> 404
docs/samples/v1beta1/lightgbm/README.md:103: /docs/samples/v1beta1/lightgbm/../../../python/lgbserver/README.md#building-your-own-ligthgbm-server-docker-image -> 404
docs/samples/v1beta1/sklearn/v2/README.md:34: https://github.com/SeldonIO/MLServer/tree/master/examples/sklearn -> 404
docs/samples/v1beta1/sklearn/v2/README.md:74: https://github.com/SeldonIO/MLServer/tree/master/examples/mms -> 404
docs/samples/v1beta1/torchserve/v2/bert/README.md:11: /docs/samples/v1beta1/torchserve/v2/bert/./sequence_classification/Transformer_handler_generalized_v2.py -> 404
docs/samples/v1beta1/transformer/feast/README.md:8: https://github.com/tedhtchang/populate_feast_online_store/blob/main/driver_stats.parquet -> 404
docs/samples/v1beta1/xgboost/README.md:44: https://github.com/SeldonIO/MLServer/tree/master/examples/xgboost -> 404
docs/samples/v1beta1/xgboost/README.md:84: https://github.com/SeldonIO/MLServer/tree/master/examples/mms -> 404
python/kserve/docs/V1alpha1ClusterServingRuntime.md:11: /python/kserve/docs/.md -> 404
python/kserve/docs/V1alpha1ModelSpec.md:8: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1alpha1ServingRuntime.md:11: /python/kserve/docs/.md -> 404
python/kserve/docs/V1alpha1TrainedModel.md:11: /python/kserve/docs/V1alpha1TrainedModelStatus.md -> 404
python/kserve/docs/V1beta1CustomExplainer.md:24: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1CustomPredictor.md:24: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1CustomTransformer.md:24: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1ExplainerSpec.md:33: https://git.k8s.io/enhancements/keps/sig-node/20190226-pod-overhead.md -> 404
python/kserve/docs/V1beta1ExplainerSpec.md:33: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1ExplainerSpec.md:37: https://git.k8s.io/enhancements/keps/sig-network/0007-pod-ready%2B%2B.md -> 404
python/kserve/docs/V1beta1ExplainerSpec.md:39: https://git.k8s.io/enhancements/keps/sig-node/runtime-class.md -> 404
python/kserve/docs/V1beta1PodSpec.md:24: https://git.k8s.io/enhancements/keps/sig-node/20190226-pod-overhead.md -> 404
python/kserve/docs/V1beta1PodSpec.md:24: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1PodSpec.md:28: https://git.k8s.io/enhancements/keps/sig-network/0007-pod-ready%2B%2B.md -> 404
python/kserve/docs/V1beta1PodSpec.md:30: https://git.k8s.io/enhancements/keps/sig-node/runtime-class.md -> 404
python/kserve/docs/V1beta1PredictorSpec.md:33: https://git.k8s.io/enhancements/keps/sig-node/20190226-pod-overhead.md -> 404
python/kserve/docs/V1beta1PredictorSpec.md:33: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1PredictorSpec.md:40: https://git.k8s.io/enhancements/keps/sig-network/0007-pod-ready%2B%2B.md -> 404
python/kserve/docs/V1beta1PredictorSpec.md:42: https://git.k8s.io/enhancements/keps/sig-node/runtime-class.md -> 404
python/kserve/docs/V1beta1TransformerSpec.md:30: https://git.k8s.io/enhancements/keps/sig-node/20190226-pod-overhead.md -> 404
python/kserve/docs/V1beta1TransformerSpec.md:30: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1TransformerSpec.md:34: https://git.k8s.io/enhancements/keps/sig-network/0007-pod-ready%2B%2B.md -> 404
python/kserve/docs/V1beta1TransformerSpec.md:36: https://git.k8s.io/enhancements/keps/sig-node/runtime-class.md -> 404

ERROR: Found 51 invalid Markdown links

```

**Comment by yuzisun** [comment]
No worries! Thanks for the great attempt ! :) We could save the report somewhere so we can later fix the remaining ones.

File: python/kserve/docs/V1beta1ExplainerSpec.md (hunk 3 above)
Comment (ckadner) [reply python/kserve/docs/V1beta1ExplainerSpec.md]: Okay. I just ran the codegen locally and the links are indeed fixed. But a few new ones don't work :-\

```YAML
Checked 1253 links (530 unique URLs) in 187 Markdown files.

README.md:3: https://coveralls.io/github/kserve/kserve?branch=master -> 404
docs/samples/README.md:36: https://github.com/kserve/kserve/tree/master/python/lightgbm -> 404
docs/samples/explanation/alibi/moviesentiment/README.md:371: /docs/samples/explanation/alibi/moviesentiment/./kfserving_text_explainer.ipynb -> 404
python/kserve/docs/V1alpha1ClusterServingRuntime.md:11: /python/kserve/docs/.md -> 404
python/kserve/docs/V1alpha1ModelSpec.md:8: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1alpha1ServingRuntime.md:11: /python/kserve/docs/.md -> 404
python/kserve/docs/V1alpha1TrainedModel.md:11: /python/kserve/docs/V1alpha1TrainedModelStatus.md -> 404
python/kserve/docs/V1beta1CustomExplainer.md:24: /python/kserve/docs/V1PodOS.md -> 404
python/kserve/docs/V1beta1CustomExplainer.md:25: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1CustomPredictor.md:24: /python/kserve/docs/V1PodOS.md -> 404
python/kserve/docs/V1beta1CustomPredictor.md:25: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1CustomTransformer.md:24: /python/kserve/docs/V1PodOS.md -> 404
python/kserve/docs/V1beta1CustomTransformer.md:25: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1ExplainerSpec.md:33: /python/kserve/docs/V1PodOS.md -> 404
python/kserve/docs/V1beta1ExplainerSpec.md:34: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1PodSpec.md:24: /python/kserve/docs/V1PodOS.md -> 404
python/kserve/docs/V1beta1PodSpec.md:25: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1PredictorSpec.md:33: /python/kserve/docs/V1PodOS.md -> 404
python/kserve/docs/V1beta1PredictorSpec.md:34: /python/kserve/docs/ResourceQuantity.md -> 404
python/kserve/docs/V1beta1TransformerSpec.md:30: /python/kserve/docs/V1PodOS.md -> 404
python/kserve/docs/V1beta1TransformerSpec.md:31: /python/kserve/docs/ResourceQuantity.md -> 404

ERROR: Found 21 invalid Markdown links

```

**Comment by ckadner** [comment]
Hi @yuzisun, I did run the codegen, which fixed some but introduced more broken links. 

I think I may want to revert those changes and only include the check-doc-links script in this PR (plus simple link fixes that are not touching generated sources or API docs). And then open a separate issue to take care of the broken links in the near future.

WDYT?

**Comment by yuzisun** [comment]
> Hi @yuzisun, I did run the codegen, which fixed some but introduced more broken links.
> 
> I think I may want to revert those changes and only include the check-doc-links script in this PR (plus simple link fixes that are not touching generated sources or API docs). And then open a separate issue to take care of the broken links in the near future.
> 
> WDYT?

I think this is fine, it needs to generated and synced anyway. We can create a separate issue for the remaining broken links.

**Comment by yuzisun** [comment]
Thanks @ckadner for the great work !

/lgtm
/approve

**Comment by ckadner** [comment]
Thank you @yuzisun for your review and merging this PR!

I have created a similar one for the KServe Website repo: https://github.com/kserve/website/pull/186 -- if/when you have time to review :-)

**Pull request merged**